client/token.go

```diff
@@ -473,10 +434,62 @@ func (impl *TokenClientImpl) GetToken(ctx context.Context, owner models.AccountA
 		Key:       tokenID,
 	}
 
-	var token Token
+	var token models.Token
 	if err := impl.client.GetTableItemByHandleAndKey(ctx, tokenStoreHandle, req, &token); err != nil {
 		return nil, fmt.Errorf("client.GetTableItemByHandleAndKey error: %v", err)
 	}
 
 	return &token, nil
 }
+
+const depositEventsField = "deposit_events"
+const ErrTableItemNotFound = "table_item_not_found"
+
+// ListAccountTokens gets all tokens of the owner by fetching all token deposit events.
+// It returns an error if the client fails to get the deposit events or the owner has
+// tokens but the client fails to get one of them.
+func (impl *TokenClientImpl) ListAccountTokens(ctx context.Context, owner models.AccountAddress) ([]models.Token, error) {
+	var tokens []models.Token
+
+	tokenIDs := make(map[models.TokenID]bool)
+
+	var start uint64 = 0
+	var limit uint64 = 100
+
+	for {
+		events, err := impl.client.GetEventsByEventHandle(
+			ctx, owner.PrefixZeroTrimmedHex(),
+			tokenStoreType, depositEventsField, start, limit,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("client.GetEventsByEventHandle error: %v", err)
+		}
+
+		if len(events) == 0 {
+			break
+		}
+
+		for _, event := range events {
+			start = uint64(event.SequenceNumber) + 1
+
+			var data models.TokenDepositEvent
+			if err := mapstructure.Decode(event.Data, &data); err != nil {
+				return nil, fmt.Errorf("mapstructure.Decode error: %v", err)
+			}
+
+			if !tokenIDs[data.ID] {
+				token, err := impl.GetToken(ctx, owner, data.ID)
+				if err != nil {
```

Comment:
In my opinion, the error should be captured and returned by GetToken() instead of here. For now, you can check error by strings.Contains(), but It would be better to handle error by checking both status code (404) and error_code field (table_item_not_found) in response. Afterward, we could modify request() function to let it return more info including status code and body.